### PR TITLE
[P2P] Remove topic from messaging - Issue #267

### LINKED ISSUE #267: [TECHDEBT] [P2P] Remove Topic from messaging
## Objective

Tending for the TODOs:

```go
./shared/modules/p2p_module.go:	Send(addr cryptoPocket.Address, msg *anypb.Any, topic debug.PocketTopic) error // TECHDEBT: get rid of topic
./shared/modules/p2p_module.go:	Broadcast(msg *anypb.Any, topic debug.PocketTopic) error                       // TECHDEBT: get rid of topic
```

## Origin Document

Tech debt

## Goals

- [ ] Remove topic from our messaging bus

## Deliverable

- [ ] Make sure that the current functionality is preserved without relying on topic

## Non-goals / Non-deliverables

- Any breaking changes in functionality apart from topic removal

## General issue deliverables

- [ ] Update the appropriate CHANGELOG
- [ ] Update any relevant READMEs (local and/or global)
- [ ] Update any relevant global documentation & references
- [ ] If applicable, update the source code tree explanation
- [ ] If applicable, add or update a state, sequence or flowchart diagram using [mermaid](https://mermaid-js.github.io/mermaid/)

## [Optional] Testing Methodology

- \__REPLACE_ME_: Make sure to update the testing methodology appropriately\_
- **Task specific tests**: `make ...`
- **All tests**: `make test_all`
- **LocalNet**: verify a `LocalNet` is still functioning correctly by following the instructions at [docs/development/README.md](https://github.com/pokt-network/pocket/tree/main/docs/development)

---

**Creator**: @deblasis 
**Co-Owners**: @Olshansk

### PR DESCRIPTION
**Pull request by deblasis**
## Description

This PR refactors our messaging and removes the string topic in favour of a more broadly used "content type + envelope" approach

## Issue

Fixes #267 

## Type of change

Please mark the relevant option(s):

- [ ] New feature, functionality or library
- [ ] Bug fix
- [x] Code health or cleanup
- [ ] Major breaking change
- [ ] Documentation
- [ ] Other 

## List of changes

- Removed topic from messaging
- Added `PocketEnvelope` as a general purpose wrapper for messages/events
- Added utility methods to `Pack` and `Unpack` messages
- Replaced the switch cases, interfaces accordingly

## Testing

- [x] `make develop_test`
- [x] [LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md) w/ all of the steps outlined in the `README`

## Required Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have tested my changes using the available tooling
- [x] I have updated the corresponding CHANGELOG

### If Applicable Checklist
- [ ] I have updated the corresponding README(s); local and/or global
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added, or updated, [mermaid.js](https://mermaid-js.github.io) diagrams in the corresponding README(s)
- [ ] I have added, or updated, documentation and [mermaid.js](https://mermaid-js.github.io) diagrams in `shared/docs/*` if I updated `shared/*`README(s)

### CODE REVIEW

File: shared/messaging/utils.go [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+func PackMessage(message proto.Message) (*PocketEnvelope, error) {
+	anyMsg, err := anypb.New(message)
+	if err != nil {
+		return nil, err
+	}
+	return &PocketEnvelope{Content: anyMsg}, nil
+}
+
+// MustPackMessage returns a *PocketEnvelope after having packed the message supplied as an argument and fails fatally in case of errors
+func MustPackMessage(message proto.Message) *PocketEnvelope {
```
Comment (andrewnguyen22) [inline shared/messaging/utils.go:21]: I'm not convinced we should have `Must` functions in our codebase. `log.Fatalf` or `panic` in non-debug code in blockchains can lead to Validators crashing or exploits. I encourage us to remove these functions if possible. 

If we decide to leave it. Then it should have a counterpart `MustUnpackMessage`

**Review by andrewnguyen22 — CHANGES_REQUESTED** [review]
Looks pretty clean with a first passthrough. Just one change

**Review by andrewnguyen22 — APPROVED** [review]
LGTM!

File: shared/messaging/proto/events.proto [hunk 2]
```diff
@@ -0,0 +1,7 @@
+syntax = "proto3";
+
+package pocket;
+
+option go_package = "github.com/pokt-network/pocket/shared/messaging";
+
+message NodeStartedEvent {}
```
Comment (Olshansk) [inline shared/messaging/proto/events.proto:7]: Thoughts on something like this?

```proto
enum NodeEvent {
	NODE_EVENT_UNDEFINED = 0;
	NODE_EVENT_STARTED = 1;
}

message NodeEvent {
     NodeEvent event = 1;
}
```

File: shared/messaging/utils.go [hunk 3]
```diff
@@ -0,0 +1,26 @@
+package messaging
```
Comment (Olshansk) [inline shared/messaging/utils.go:1]: Though I'm also at fault of always using `utils.go`, I think `envelope.go` could help scope from becoming a catchall.

Weakly held point so lmk what you think.

File: shared/messaging/utils.go [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+func PackMessage(message proto.Message) (*PocketEnvelope, error) {
+	anyMsg, err := anypb.New(message)
+	if err != nil {
+		return nil, err
+	}
+	return &PocketEnvelope{Content: anyMsg}, nil
+}
+
+// UnpackMessage extracts the message inside the PocketEnvelope decorating it with typing information
+func UnpackMessage(envelope *PocketEnvelope) (protoreflect.ProtoMessage, error) {
```
Comment (Olshansk) [inline shared/messaging/utils.go:19]: Everyone who calls this function will need to return a protobuf of a specific type (as I see in the usage and in the text).

What do you think of leveraging generics?

```go
func UnpackMessage[T protoreflect.ProtoMessage] (T, error) {... }
```

**Review by Olshansk — COMMENTED** [review]
@deblasis I know this was already merged, but got a couple nits and one suggestion while I was rebasing over the changes. PTAL

File: shared/messaging/utils.go (hunk 4 above)
Comment (deblasis) [reply shared/messaging/utils.go:19]: Nice catch!

File: shared/messaging/proto/events.proto (hunk 2 above)
Comment (deblasis) [reply shared/messaging/proto/events.proto:7]: I'd prefer individual events that have their own properties and/or enums if needed. 

This is just a "flagging event" with no properties so I get your point, however let's think about more complex events.

Lemme try to explain what I mean:

If we have a "catch-all" NodeEvent, we'd have to inspect the content on the handler side everytime, also you would end up having properties in there that make sense for an event but not for another. 
This increases the risk of breaking changes down the line and also complexity IMHO.

With individual types, instead, you just look at the envelope if that makes sense and you treat apples like apples and oranges like oranges.

File: shared/messaging/messages.go [hunk 5]
```diff
@@ -0,0 +1,11 @@
+package messaging
+
+import "strings"
+
+const (
+	DebugMessageEventType = "pocket.DebugMessage"
+)
+
+func (x *PocketEnvelope) GetContentType() string {
```
[the remark this answers is not in the document]
Comment (deblasis) [reply shared/messaging/messages.go:9]: No, I misnamed it

**Comment by deblasis** [comment]
Followup PR for NITS: https://github.com/pokt-network/pocket/pull/359

**Pull request merged**